pulpcore/cli/ansible/remote.py

```diff
@@ -0,0 +1,100 @@
+import click
+
+from typing import Any, Dict, Iterator
+
+from pulpcore.cli.common.generic import (
+    list_entities,
+    show_by_name,
+    destroy_by_name,
+)
+from pulpcore.cli.common.context import (
+    PulpContext,
+    pass_pulp_context,
+    pass_entity_context,
+    PulpEntityContext,
+)
+from pulpcore.cli.ansible.context import (
+    PulpAnsibleRemoteRoleContext,
+    PulpAnsibleRemoteCollectionContext,
+)
+
+
+@click.group()
+@click.option(
+    "-t",
+    "--type",
+    "remote_type",
+    type=click.Choice(["role", "collection"], case_sensitive=False),
+    default="role",
+)
+@pass_pulp_context
+@click.pass_context
+def remote(ctx: click.Context, pulp_ctx: PulpContext, remote_type: str) -> None:
+    if remote_type == "role":
+        ctx.obj = PulpAnsibleRemoteRoleContext(pulp_ctx)
+    elif remote_type == "collection":
+        ctx.obj = PulpAnsibleRemoteCollectionContext(pulp_ctx)
+    else:
+        raise NotImplementedError()
+
+
+remote.add_command(list_entities)
+remote.add_command(show_by_name)
+remote.add_command(destroy_by_name)
+
+
+@remote.command(
+    context_settings=dict(
+        ignore_unknown_options=True,
+    )
+)
+@click.option("--name", required=True)
+@click.option("--url", required=True)
+@click.argument("args", nargs=-1, type=click.UNPROCESSED)
+@pass_entity_context
+@pass_pulp_context
+def create(
+    pulp_ctx: PulpContext,
+    remote_ctx: PulpEntityContext,
+    name: str,
+    url: str,
+    args: Any,
+) -> None:
+    """Use '-t collection' after remote to create a collection remote"""
+    body: Dict[str, str] = {"name": name, "url": url}
+    if args:
+        body.update(extra_options(args))
+    result = remote_ctx.create(body=body)
+    pulp_ctx.output_result(result)
+
+
+@remote.command(
+    context_settings=dict(
+        ignore_unknown_options=True,
+    )
+)
+@click.option("--href", required=True)
+@click.argument("args", nargs=-1, type=click.UNPROCESSED)
+@pass_entity_context
+@pass_pulp_context
+def update(pulp_ctx: PulpContext, remote_ctx: PulpEntityContext, href: str, args: Any) -> None:
+    """ "
+    Performs a partial update of the remote
+    The one annoying part of this is that you want update a collection remote you need to
+    specify with -t parameter after 'remote' or else defaults to role remote
+    """
+    body: Dict[str, Any] = dict()
+    if args:
+        body.update(extra_options(args))
```

Comment:
interesting Idea! I need to think more about this, as you already realized, it lacks the click validation and help automation. Maybe we can get it with a bit of click.Context hackery.

Reply:
I think, I am onto something here:
https://github.com/pulp/pulp-cli/pull/64
It's even less hacky than i thougt.
Can you see how it maps to your problem?